### PR TITLE
debug: add usage guide for function tracing of user TAs

### PR DESCRIPTION
**Pull request by b49020**
Add ftrace user guide for function tracing of user Trusted Applications
and corresponding user libraries.

Depends on https://github.com/OP-TEE/optee_os/pull/2912 and https://github.com/OP-TEE/optee_client/pull/155.

### CODE REVIEW

File: debug/gprof.rst [hunk 1]
```diff
@@ -14,7 +14,7 @@ Usage
 *****
 
     - Build OP-TEE OS with ``CFG_TA_GPROF_SUPPORT=y``. You may also set
-      ``CFG_ULIBS_GPROF=y`` to instrument the user TA libraries (libutee,
+      ``CFG_ULIBS_INTRUMENT=y`` to instrument the user TA libraries (libutee,
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/gprof.rst]: Its `INSTRUMENT`, will correct.

File: debug/index.rst [hunk 2]
```diff
@@ -7,3 +7,4 @@ Debugging techniques
    abort_dumps
    benchmark
    gprof
+   ftrace
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/index.rst]: Ok.

File: debug/ftrace.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+compiled with ``-pg``. Once collected, the function graph data is formatted
+in the ``ftrace.out`` format and sent to ``tee-supplicant`` via RPC, so they
+can be saved to disk, later processed and displayed using helper script called
+``symbolize.py`` present as part of ``optee_os`` repo.
+
+Usage
+*****
+
+    - Build OP-TEE OS and OP-TEE Client with ``CFG_TA_FTRACE_SUPPORT=y``. You
+      may also set ``CFG_ULIBS_INTRUMENT=y`` in OP-TEE OS to instrument the
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/ftrace.rst]: INSTRUMENT

File: debug/ftrace.rst [hunk 4]
```diff
@@ -0,0 +1,36 @@
+.. _ftrace:
+
+Ftrace
+######
+This section describes to generate function call graph for user Trusted
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/ftrace.rst]: ok

File: debug/ftrace.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+Usage
+*****
+
+    - Build OP-TEE OS and OP-TEE Client with ``CFG_TA_FTRACE_SUPPORT=y``. You
+      may also set ``CFG_ULIBS_INTRUMENT=y`` in OP-TEE OS to instrument the
+      user TA libraries (libutee, libutils, libmpa).
+
+    - Build user TAs with ``-pg``, for instance using: ``CFLAGS_ta_arm64=-pg``.
+      Note that instrumented TAs have a larger ``.bss`` section. The memory
+      overhead depends on FTRACE_BUF_SIZE macro which can be configured
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/ftrace.rst]: you mean CFG_ULIBS_INSTRUMENT right? Will correct.

File: debug/ftrace.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      specific to user TAs like: ``CPPFLAGS_ta_arm64 = -DFTRACE_BUF_SIZE=2048``
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/ftrace.rst]: ok

**Comment by b49020** [comment]
Updated PR to incorporate comments.

File: debug/ftrace.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply debug/ftrace.rst]: Sorry, bad copy/paste. I meant add ` `` ` around `FTRACE_BUF_SIZE`  for formatting purpose.
_not around `CFG_ULIBS_INTRUMENT` :(_

[the remark this answers is not in the document]
Comment (b49020) [reply debug/ftrace.rst]: ok

File: debug/ftrace.rst [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+Usage
+*****
+
+    - Build OP-TEE OS and OP-TEE Client with ``CFG_TA_FTRACE_SUPPORT=y``. You
+      may also set ``CFG_ULIBS_INSTRUMENT=y`` in OP-TEE OS to instrument the
+      user TA libraries (libutee, libutils, libmpa).
+
+    - Build user TAs with ``-pg``, for instance using: ``CFLAGS_ta_arm64=-pg``.
+      Note that instrumented TAs have a larger ``.bss`` section. The memory
+      overhead depends on FTRACE_BUF_SIZE macro which can be configured
```
Comment (etienne-lms) [inline debug/ftrace.rst]: > s/CFG_ULIBS_INTRUMENT/\``CFG_ULIBS_INTRUMENT\``/

Sorry, bad copy/paste. I meant s/FTRACE_BUF_SIZE/\`\`FTRACE_BUF_SIZE\`\`/:
Add \`\` around `FTRACE_BUF_SIZE` for formatting purpose,  _(not around `CFG_ULIBS_INTRUMENT` :()_

Comment (b49020) [reply debug/ftrace.rst]: Ok, will add `` around FTRACE_BUF_SIZE for formatting purpose.

File: debug/ftrace.rst [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+Ftrace
+######
+This section describes how to generate function call graph for user Trusted
+Applications using ``ftrace``.
+
+The configuration option ``CFG_TA_FTRACE_SUPPORT=y`` enables OP-TEE to collect
+function graph information from Trusted Applications running in user mode and
+compiled with ``-pg``. Once collected, the function graph data is formatted
+in the ``ftrace.out`` format and sent to ``tee-supplicant`` via RPC, so they
```
[the remark this answers is not in the document]
Comment (b49020) [reply debug/ftrace.rst]: Yes this doc is inspired from gprof stuff. I am not sure what information will user get from `ftrace.out` format as that is consumed by `symbolize.py` script. IMO, we can add example function graph output using `symbolize.py` script, your views?

**Comment by b49020** [comment]
Updated PR to incorporate recent changes. Also, update gprof doc to use `CFG_TA_MCOUNT`.

**Comment by b49020** [comment]
Tag applied.

**Pull request merged**